src/dtx/dtx_srv.c

```diff
@@ -88,11 +116,24 @@ dtx_handler(crt_rpc_t *rpc)
 	case DTX_CHECK:
 		/* Currently, only support to check single DTX state. */
 		if (din->di_dtx_array.ca_count != 1)
-			rc = -DER_PROTO;
-		else
-			rc = vos_dtx_check(cont->sc_hdl,
-					   din->di_dtx_array.ca_arrays,
-					   NULL, NULL, false);
+			D_GOTO(out, rc = -DER_PROTO);
+
+		/* If the DTX on this server is for read only operations,
+		 * then di_epoch will be non-zero. If such epoch is older
+		 * than current server guard epoch, regard it as prepared
+		 * in spite of whether it has been really handled on this
+		 * server or not before current server the latest restart
+		 * because the guard epoch on this server will refuse any
+		 * modification with old eppch, that have the same effect
+		 * as related read only operations for this DTX have been
+		 * executed on current server.
+		 */
+		if (din->di_epoch != 0 &&
+		    din->di_epoch <= dtx_get_guard_epoch())
+			D_GOTO(out, rc = DTX_ST_PREPARED);
```

Comment:
Is this the right return code?  Should it be DER_TX_RESTART?

Reply:
DTX_ST_PREPARED is the expected return value. As described in the comment, this check will happen after current server restart, because of the restart, the former cached DTX information for read only operations on current server are all lost, then if some other server (leader) checks whether some read only operations for related DTX have ever been executed before current server restart, current server will regard it as "executed".

Current server cannot reply "-DER_TX_RESTART", because related DTX may have been committable on the leader before current server restart, its original sponsor (the client and application) may have already exited.